Improve CI workflows

LecrisUT opened the pull request:
- [x] Add concurrency. This helps cancel ongoing jobs when you push a new commit
- [x] Unify the CI jobs to one file
- [x] Migrate strategy matrix to a separate file
- [ ] ~~Group matrix jobs. Dunno how to do this, might postpone~~ Partially organized in Checks view
- [x] Cleanup ci jobs. E.g. `submodules` is not needed
- [x] Setup Swiglang from upstream GitHub Actions. See `.github/actions/setup-lang/action.yaml`

After some experimentation, I propose a clean solution using `fmf`

<details><summary>What's fmf?</summary>

[FMF](https://fmf.readthedocs.io/en/stable/features.html#features) in a nutshell is an extension to yaml such that you can define and combine yaml files hierarchically. E.g. in a top-level file (`/Lib/main.fmf`) you have
```yaml
github-matrix:
  version: [ "" ]
  cpp-std: [ "c++11","c++14","c++17" ]
```
then in a child file (`/Lib/python/main.fmf`) you can write as:
```yaml
github-matrix+:
  version: [ "3.8","3.9","3.10","3.11","3.12" ]
```
which is equivalent to:
```yaml
github-matrix:
  version: [ "3.8","3.9","3.10","3.11","3.12" ]
  cpp-std: [ "c++11","c++14","c++17" ]
```

Primarily `fmf` is used in [`tmt`](https://tmt.readthedocs.io/en/stable/) which is a testing-framework where you can define test hierarchically. It might be useful to also use that infrastructure, but for now, one step at a time.

Some tips of how to view the data:
```console
## Install fmf
$ pip install fmf
## Go to the root of fmf (the folder that contains `.fmf/version`)
$ cd Lib
## Show all leaf nodes
$ fmf ls
## Show all data in each node
$ fmf show
## Show only the data in python. Note it is a regex expression so `java`/`javascript` needs more sanitization
$ fmf show --name=/python
```

</details>


<details><summary>Why fmf?</summary>

Originally I was planning to store these as json files, but it soon proved quite unwieldy especially due to the duplications across the variants. This approach will allow:
- reduce duplication and maintenance. E.g. if `c++20` should be added to all languages as experimental, this can be done in the top-level `/Lib/main.fmf` and all languages will inherit it.
- simplify the sanitization of variables. Passing strings and interpreting them back as json for `strategy.matrix` proved quite complicated adding too much boilerplate to the workflows making them unnavigable. In this approach, the whole `matrix-github` is passed as a single object.

</details>

Upstream issues:
- [x] https://github.com/tj-actions/changed-files/issues/1528

ojwb commented on .github/workflows/nuget.yml:
```diff
@@ -2,19 +2,21 @@ name: Windows Nuget Build
       - 'appveyor.yml'
 
-permissions:
-  contents: read  #  to fetch code (actions/checkout)
```
This was deliberately added by 62e27bb61680482ec9a1a0a27f7bb29946c9472e.

It seems this is just setting the default, but I assume the idea was it's better to explicitly set it to the value we want rather than rely on the default.  I suggest you should discuss with @sashashura as this part is effectively reverting their changes.

LecrisUT replied:
Oh, I've had a read and I've got the permission mixed up. I thought it was additive, not exclusive. I'll revert and add minimal comments for it

ojwb commented:
> * `path` [default=/Lib/${input.lang}]: A variable to filter the paths where the CI should run or not

Note this is more complicated to do correctly that just `/Lib/${input.lang}` - that wouldn't be useful even for PRs as we want to know if a PR to fix e.g. a Python-specific issue breaks any other target languages before merging it.  So anything touching generic library, testsuite or source files needs to trigger a full set of jobs, but we certainly can save running jobs when the only files touched are target-language-specific such as `Lib/python`, `Examples/python`, `Examples/test-suite/python` and `Source/Modules/python.cxx`.

> * `ignore-path`: ignore the arguments in `path` and run the workflow. This is to be used on `master` branch

If we've got accurate filtering of which jobs to run (which would be nice), I don't really see a good enough reason to avoid using it on master too.  We're not going to have long gaps without a full set of jobs because anything which touches a generic files needs to trigger jobs for all languages.

Hooking up `on.workflow_dispatch` such that a full set of jobs can be manually triggered from the web UI might be handy sometimes - e.g. I can imagine wsfulton might want to be able to verify CI is definitely all green before making a release.

LecrisUT commented:
> So anything touching generic library, testsuite or source files needs to trigger a full set of jobs, but we certainly can save running jobs when the only files touched are target-language-specific such as `Lib/python`, `Examples/python`, `Examples/test-suite/python` and `Source/Modules/python.cxx`.

That's precisely the design that I was planning to add. My plan is to use the design [here](https://github.com/pypa/build/blob/main/.github/workflows/reusable-change-detection.yml) and a json/yaml file configuration to define all the languages, variants and paths.

> If we've got accurate filtering of which jobs to run (which would be nice), I don't really see a good enough reason to avoid using it on master too. We're not going to have long gaps without a full set of jobs because anything which touches a generic files needs to trigger jobs for all languages.

My reasoning for that is to have a workflow that could do a sanity check (e.g. there are changes in python-dev, or changes in the compiler, etc.). Pushes to master should be fine in terms of frequency, but this could be on a weekly schedule as well.

> Hooking up on.workflow_dispatch such that a full set of jobs can be manually triggered from the web UI might be handy sometimes

👍

ojwb commented:
> My reasoning for that is to have a workflow that could do a sanity check (e.g. there are changes in python-dev, or changes in the compiler, etc.). Pushes to master should be fine in terms of frequency, but this could be on a weekly schedule as well.

In the case of `python-dev` that comes from the Ubuntu package repos which means it'll only change for targetted fixes for security or other high severity bugs, or else when we move to a new Ubuntu release but for the latter case the CI config file changes (from saying e.g. `ubuntu-20.04` to `ubuntu-22.04`) and that changing should trigger a full set of jobs anyway.

There are some dependencies that get installed from sources with more liberal update policies than Ubuntu, but I think the benefits of reducing the number of CI jobs for many pushes to master outweighs the potential downside of not seeing the effects of a breaking change to a target language dependency until we next happen to trigger a build for it (that's actually true already, it's just that such triggering would happen a bit less often with these changes).

LecrisUT commented:
By python-dev I was referring to python-3.qw-dev from setup-python github action. And similarly for the other languages, it should be designed to allow for running the latest development versions so that breaking changes can be caught early in the development stage instead of after release

LecrisUT commented:
@ojwb I have added an implementation that simplifies the CI runs and has selective runs. Could you get someone to take a look at it and discuss it? I did drop a lot of the variants and it would be good to have someone look at which variants are missing and need to be included/fixed.

erezgeva commented:
Hi Cristian,

I like your approach.
It looks in the right direction.

In my draft, I was looking to do:

- Run the affected language only, for example if I change a test in `Examples/test-suite/php/`, I want only the php tests running and not the rest of the languages.
- Move the windows test from appveyor to github.

But back to your request, regarding the `autotools`.
Can you please point the error or the location you wish me to have a look?

Thanks for your work, Erez Geva

LecrisUT commented:
>     * Move the windows test from appveyor to github.

If it's ok with just checking the default msvc, I can add those toolchains as well. I actually have a workflow that runs for the latest fedora release toolchains, windows and macos. It doesn't cover variants of gcc and llvm, but maybe we can do that by mixing the workflows

> But back to your request, regarding the `autotools`. Can you please point the error or the location you wish me to have a look?

Well for that there are missing environment variables. Can you evaluate using `fmf show` to see what variants and what environment variables are currently missing? Basically the `github-matrix` shows what will be replaced in `strategy.matrix` and you should be able to confirm with the running wokflows that the variant is there. We might need more `include` to add specific flags to each one.

Also, there are failing tests with `gcc-13` which should be due to the manual auto-tools script. Try to look at the failing jobs to check what environment variable is missing.

Note that in the #2646 there I am doing the cmake workflows rework, which can greatly simplify over the autotools action

erezgeva commented on .github/workflows/step_change-detection.yml:
```diff
@@ -0,0 +1,60 @@
+            Lib/std/**
+            Lib/typemaps/**
+            Lib/xml/**
+            Lib/*
```
Should also probe changes in (which require to run all tests on all langauges)
Examples/*
Examples/test-suite/*

erezgeva commented on .github/workflows/step_change-detection.yml:
```diff
@@ -0,0 +1,60 @@
+            .fmf/*
+            std/**
+            typemaps/**
+            xml/**
```
Should also probe changes in per language.
Examples/*
Examples/test-suite/*/**
There is one annoying exception, the schemerunme folder is used by guile.
It was probably planned as a shared folder for scheme based languages.

LecrisUT replied:
This one is a bit tricky because I was using the path of the library to get the language path. I'll try to convert it to a separate job and get the info from fmf.

LecrisUT replied:
Ok, found a suitable workflow for that. Only `schemerunme` has this special treatment?

erezgeva replied:
> Ok, found a suitable workflow for that. Only schemerunme has this special treatment?

As I check all languages in `Examples/test-suite`, it is the only one at the moment.

erezgeva commented:
Comparing old with new.
I do not see all variable from environment.
`    env:`
`      SWIGLANG: ${{ matrix.SWIGLANG }}`
`      PY2: ${{ matrix.PY2 }}`
`      VER: ${{ matrix.VER }}`
`      ENGINE: ${{ matrix.ENGINE }}`
`      SWIG_FEATURES: ${{ matrix.SWIG_FEATURES }}`
`      GCC: ${{ matrix.GCC }}`
`      CSTD: ${{ matrix.CSTD }}`
`      CPPSTD: ${{ matrix.CPPSTD }}`
`      CPPFLAGS: ${{ matrix.CPPFLAGS }}`

Make file uses both macros from command line, but may also use environment variables.
As far as I remember, the environment variables are forward also to all application invoked on `run` stage (of the same scope).

Not sure if we need all the variables, but some are probably missed :-)
I guess for example we do not need PY2, as we use python3.

LecrisUT commented:
NOTE: javascript is not running in this one yet. How is that one supposed to work? Also how to deal with `Lib/xml`, `Examples/android`?

erezgeva commented:
> > ~ To make my debugging quicker, I usually limits the tests I run. How do you do that here? For example run only Ruby tests? ~
> 
> I take it that you found the dispatch workflow? You do need to have it in `main` branch for this to be trigger-able.
> 
> > Comparing old with new. I do not see all variable from environment.
> 
> Indeed that's where I could use some help. E.g.:
> 
> ```yaml
>         - SWIGLANG: go
>           VER: '1.12'
>           CSTD: gnu11
>         - SWIGLANG: go
>           VER: '1.17'
> ```
> 
> This one would indicate that `1.17` should still run properly, but only `1.6` and `1.8` variants were successful. The base ubuntu was shifted, but I wouldn't think that that would cause the issue like that.
> 
> Could you help and detangle:
> 
> * which environment variable is for base swig runner, and which are for `SWIGLANG`
> * Which environment variable is specific for only one language
> * Which variables are missing in the current runs that should be added.
> 
> > * GCC 13 (as it is experimental, we can skip it on first iterator)
> 
> Experimental flag was created by me, it was not in the original CI. I've added it because CI were failing and we needed the other runs to run for now.

I'll have a look on week end.

Is there a way to add more variables in the fmf file?
So we can add the variant variables like `CSTD`.
For example in `Lib/ruby/main.fmf`?
Of course we will need to fetch them in `step_get-variants.yml`.

LecrisUT commented:
> Is there a way to add more variables in the fmf file? So we can add the variant variables like `CSTD`. For example in `Lib/ruby/main.fmf`? Of course we will need to fetch them in `step_get-variants.yml`.

Add them like a normal yaml content, specifically like the content you would add to the github action `strategy.matrix`, e.g.:
```yaml
github-matrix+:
  version: [ "2.0","2.1","2.2","2.3","2.4","2.5","2.6","2.7","3.0", "3.1","3.2" ]
    include:
      - version: "3.2"
        cstd: "11"
```
Adds `cstd: 11` to all ruby version variants `3.2`. After that just need to add the equivalent `env.CSTD: ${{ matrix.cstd }}` or equivalent to consume the variable properly.

erezgeva commented:
> NOTE: javascript is not running in this one yet. How is that one supposed to work? Also how to deal with `Lib/xml`, `Examples/android`?

There are 4 variants of javascript: jsc, napi, v8 and node.
During `configure` the 4 variants are probed and could be all available.
They are tested separately. Each variant uses an `ENGINE` variable.
and uses a different library or a different application (which the`configure` found).
On the CI, v8 is currently not used.
As node does not have a library folder, we could do something like the following in the javascript `main.fmf`:

    github-matrix+:
      version: [ "18", "20",  ...]
        include:
          - version: "18"
            cstd: "c++17"
            gcc: "13"
            engine: [ "node" ]
          - version: "20"
            cstd: "c++20"
            gcc: "13"
            engine: ["node", "napi" ]`


Android uses Java (although Google recommends Kotlin),
But require Android SDK and NDK.
I do not think we test them in the CI.

The XML is a generator, not a real language. 

    Much of SWIG's current parser design was originally motivated
     by interest in using XML to represent SWIG parse trees.
    Although XML is not currently used in any direct manner, the parse tree structure,
     use of node tags, attributes, and attribute namespaces
     are all influenced by aspects of XML parsing.
    Therefore, in trying to understand SWIG's internal data structures,
     it may be useful to keep XML in the back of your mind as a model.

I do not think we test it in CI.

@ojwb are Android and XML tested in CI?

erezgeva commented:
> Can you walk me through this a bit more when you have time? I need more context for #2646.
> 
> * About `engine: node` which `Lib` files does it listen to?

    The more recent -napi generator produces node-addon-api that interfaces to Node.js through Node-API. 
'node` uses `napi`.

> * Shouldn't some of the `javascript` files be common and have some specific folders overwrite the default?

`/Lib/javascript` is empty. The 3 libraries are not compatible.
v8 and jsc uses C++, while node uses C.

> * How do you specify the engine on the `swig` cli? Also in the test `Makefile` how is it invoked?

You set the value of `ENGINE` to `node`, `jsc` or `v8`. You can set the variable in the environment, as make use the environment variables by default.

> * What are the preparation steps for each language? I will convert them to native Github Actions, and for readability I can separate them to `.github/actions/get_language/actions.yaml`

What you see in the old CI.
We call Tools/GHA-linux-install.sh -> Tools/CI-linux-install.sh to install the proper Debian packages on Ubuntu. 
Than we call Tools/CI-linux-environment.sh to set the environment for the specific language.
Than calling `configure`.
Than calling 4 times the  make, which you copy.
Some of the parameters passing to the make file are done through environment variables in addition to the command line.  Usually the environment variables remain the same cross the 4 times you invoke the make.

In assent, I think  you copy most. I think we just need to add the missing environment variables and add the variant, on the way you already suggest.
I think we are on the right path.

> * How is the `GCC` used for language specific tests. IMO the language tests should not be compiler version dependent, unless there is a way to print and trigger the specific combination like with `-c++`

Some test C++ and some uses C.
The GCC usually specify the GCC version.

> * How are the `cpp-std` used in both the test and in the swig itself?

Most of the parameters/variables are used during the tests, so they are used in `Examples/Makefile`, `Examples/test-suite/common.mk`
The build of swig itself, usually use the configuration results, and do not use environment variables or command line variables.

Regarding `CPPSTD`.
During test the swig generate C or C++ code. The test then compile the swig generate code.
The GCC set the compiler version to use on the compilation (probably also of the swig tool itself?).
The `CPPSTD` add `--std=` to set the C/C++ standard to use on the compilation.
We want to test different C/C++ standards, to make sure we support new standards and keep backward compatible with old standards.

LecrisUT commented:
Thanks for walking me through this. I have some ideas of how to simplify this on the cmake side. I can basically consolidate all of the scripts except for the language package installations to a single execution in the github workflow. I think I can even simplify it so that only one swig build is run and then all the language tests run the same executable.

For this one though I will fix the environment issues and try to recreate the missing configurations.

LecrisUT commented:
@ojwb @erezgeva I think this variant should address most issues and still be relatively concise.
Let me try and give a short summary of what I've changed in the behaviour of the CI:
- All language variants are extracted from `fmf` metadata in `/Lib`. You can run `fmf show --path ./Lib` to get a sense of the CI variants that are generated
- Added conditional run for CIs depending on what files are changed `base` runs all, while changes in `Lib`/`Examples` only run specific languages (except if base example files are changed of course)
- Consolidated the language variant tests:
  - All languages test for `c++11`, `c++14`, `c++17` (and `c++20` experimental)
  - All languages check for `c11`/`c17` standards
  - Running on `ubuntu-22.04` for all languages. I would even go as far to change it to `ubuntu-latest`
    If necessary for older languages, back-port the older `ubunutu-20.04` as needed.
- Removed some variants. Please check if there are any missing ones that should still be added.
  - python `2`, `3.5`, `3.6`, `3.7` since those are out of support
  - Not familiar with the other languages that need to be deprecated
- Removed toolchain tests for `gcc-7`, `gcc-8`, `gcc-9`. According to `pkgs.org` RHEL9, Debian 11, Ubuntu 22.04 all have `gcc-10` or higher. Worth to gradually phase out those variants.


Some comments:
- I don't think we should encourage users to use non standard, and in 2023 there should not be an issue enforcing that minimum standard. I thought this was also documented, but the site is down so I can't check that.
- There are still many variants that should be deprecated out like `node: 12`
- What's the deal with `-builtin -O`, `-builtin` and `-DDOH_POISON` runners?

erezgeva commented:
Hi Cristian,

I was looking on Ruby testing.
I compare Ruby 2.3
What I see is on old testing:
configure is: `--disable-cpp11-testing --without-alllang --with-ruby`
With new: `'CXXFLAGS=-std=c++11 ' 'CFLAGS=-std=c11 ' --without-alllang --with-ruby`

The CFLAGS/CXXFLAGS are not needed on configuration, but are harmless.
But the reason the test fails, is because it do not disable the CPP 11, which probably fails.
We need to set the proper configuration and mark each test, is it a C or a C++ test :-)

I have another thought.
For now we do a full configure-build-test cycle in each test.
But we can use artifices and slit the 3 parts.
For example during configuration,
These flags "--without-alllang --with-ruby" do not really change the test.
Only flags like "--disable-cpp11-testing".

The split may reduce testing time, even when testing a single language.
We have many repetitions in our testing.

Erez

LecrisUT commented:
Thanks @erezgeva, I have added an overload `matrix.config-opts` where we can add `--disable-cpp11-testing` to. Could you help me confirm which tests need those to be added. Other `ruby` tests fail at `Install dependencies`, which could be because of using `ubuntu-22.04`. Could you help me add any relevant flags that are needed like they were done in f9926ea1da3448d3a2328e9ad3f7402118eb0359?

> I have another thought.
> For now we do a full configure-build-test cycle in each test.
> But we can use artifices and slit the 3 parts.
> For example during configuration,
> These flags "--without-alllang --with-ruby" do not really change the test.
> Only flags like "--disable-cpp11-testing".

IMO, the whole script is a mess right now, but I am not fluent enough with autotools to clean it up. Could we postpone this for future PR so we can get this one merged soon? I have plans to simplify those points using cmake anyway.

> But the reason the test fails, is because it do not disable the CPP 11, which probably fails.
> We need to set the proper configuration and mark each test, is it a C or a C++ test :-)

Agree, but the current test-suite makes it a bit hard to filter those out. Lets move it to after we get a cleaner test-suite output. I'll see how it looks in the ctest side.

erezgeva commented:
I run the 107 jobs on the old tests.

Attached are the 107 configurations.
[cfg_107.log](https://github.com/swig/swig/files/12677685/cfg_107.log)

And 98 make - running the language tests.
[mk_98.log](https://github.com/swig/swig/files/12677686/mk_98.log)

I focus on the configuration/parameters values.

After running the new tests, I will compare the two.

LecrisUT commented:
@erezgeva I have managed to back-port almost all configurations. If you look at the CI, most failed ones are marked as `experimental` and the failing ones have a non-std configuration that passes (these should be fased out imo). The only remaining ones are guile and javascript/napi. The former I don't know how to port to a proper GH action, but the latter it should be doable by using setup-node action.

erezgeva commented:
> @erezgeva I have managed to back-port almost all configurations. If you look at the CI, most failed ones are marked as `experimental` and the failing ones have a non-std configuration that passes (these should be fased out imo). The only remaining ones are guile and javascript/napi. The former I don't know how to port to a proper GH action, but the latter it should be doable by using setup-node action.

I try to test, but nothing happens in my action.
See: https://github.com/erezgeva/swig/actions/runs/6291692604/job/17080318287

LecrisUT commented:
That is because the changed-files detection running on the same branch is checked against the previous commit, and I did not set it that github actions should trigger these. Will do so when the structure is reviewed.

You can trigger it either by opening a PR on your fork against `master` (because it changes base files) or fast-forward `master` and trigger the workflow manually

erezgeva commented:
> That is because the changed-files detection running on the same branch is checked against the previous commit, and I did not set it that github actions should trigger these. Will do so when the structure is reviewed.
> 
> You can trigger it either by opening a PR on your fork against `master` (because it changes base files) or fast-forward `master` and trigger the workflow manually

According to https://docs.github.com/en/actions/using-workflows/manually-running-a-workflow
"To trigger the workflow_dispatch event, your workflow must be in the default branch."

I just need to mark this branch as default (the default branch is master until you change).

Erez

erezgeva commented:
I think the job names are too long,
GitHub tent to cut long string. These of us which need bigger text, the cut is worse.

Jobs were named like: "ruby 3.1 gcc c++11".
Only changed information is used in jobs name.

The work "run" does not add any information. All jobs are running.
Also "test" does not add any information, as all jobs are tests!

We can fetch the action with GitHub CLI/ REST API, and read the full text.
But using the GUI is more convenient.

Thanks
   Erez

LecrisUT commented:
> We can fetch the action with GitHub CLI/ REST API, and read the full text.
> But using the GUI is more convenient.

Note that if you navigate: `Checks` -> `CI on: pull_request` you have a much better view of the jobs, especially the `pass` job which takes into account all `(exeprimental)` jobs.

LecrisUT commented:
@ojwb @erezgeva @wsfulton Final review please. All passing tests have been backported. Please see the `pass` jobs for individual languages and the `pass` for the whole CI. These jobs will fail if a non-`experimental` job is failing, but will not fail if those are skipped. Other failing jobs are marked as `experimental`.

LecrisUT commented:
@ojwb @erezgeva @wsfulton gentle poke to request some attention on this PR

erezgeva commented:
Hi @LecrisUT ,

I like your work and I think the pull request is a good one.

From my side there are 2 blocking:
1. The job title on the web are too long and contain words like 'test' 'language' and 'run'. Theses worlds provide not information. The words appear on the start of the title make it harder to read.
**Please use the original titles!**
See my run: https://github.com/erezgeva/swig/actions/runs/6357334511

2. There are jobs that fails. They are probably new tests.
 I do not oppose adding extra tests for a "full" testing.
 **But the default tests must pass!**

LecrisUT commented:
>     1. The job title on the web are too long and contain words like 'test' 'language' and 'run'. Theses worlds provide no information. The words appear on the start of the title make it harder to read.
>        **Please use the original titles!**
>        See my run: https://github.com/erezgeva/swig/actions/runs/6357334511

Agree. After some experimentation I found where the naming comes from and how to simplify, but I cannot get rid of the folder-like structure though, i.e. the `CI/.../...`. So, how should we:
- distinguish between `lang` and `toolchain` workflows? I think the best would be with emoji like 🛠️ , but what can we use for lang? 🗨️ ?
- are the current versions still too long? Can you be more specific on how you would want them simplified?

>     2. There are jobs that fails. They are probably new tests.
>        I do not oppose adding extra tests for a "full" testing.
>        **But the default tests must pass!**

Pay attention to the CI/pass. There it shows if the normal tests fail or not. The additional tests are for `experimental` to highlight and keep track of how things are failing there and what needs to be done. What I can do is to mask the failing tests when pushed to `main`, that way the checks appear green in the commit history. I strongly recommend to keep them and show the failures in the PR.

erezgeva commented:
> > ```
> > 1. The job title on the web are too long and contain words like 'test' 'language' and 'run'. Theses worlds provide no information. The words appear on the start of the title make it harder to read.
> >    **Please use the original titles!**
> >    See my run: https://github.com/erezgeva/swig/actions/runs/6357334511
> > ```
> 
> Agree. After some experimentation I found where the naming comes from and how to simplify, but I cannot get rid of the folder-like structure though, i.e. the `CI/.../...`. So, how should we:
> 

I like the folder like view.

> * distinguish between `lang` and `toolchain` workflows? I think the best would be with emoji like 🛠️ , but what can we use for lang? 🗨️ ?

The folder make it easier to read for example:

` TCL lang >`
`           C++ 11`
`           C++ 17`

You can put the 'lang' and 'tool' after as most developers know TCL is a language and 'GCC' a compiler (tool)
i.e. 
`GCC tool`  
`TCL lang`

As for emoji, I like the idea. You're choosing is better than my ideas.


> * are the current versions still too long? Can you be more specific on how you would want them simplified?


> 
> > ```
> > 2. There are jobs that fails. They are probably new tests.
> >    I do not oppose adding extra tests for a "full" testing.
> >    **But the default tests must pass!**
> > ```
> 
> Pay attention to the CI/pass. There it shows if the normal tests fail or not. The additional tests are for `experimental` to highlight and keep track of how things are failing there and what needs to be done. What I can do is to mask the failing tests when pushed to `main`, that way the checks appear green in the commit history. I strongly recommend to keep them and show the failures in the PR.

There is no difference.
If the PR fails, it will **not** be merged into 'main'.
Developers can not differ the 2 kinds.

What we can do is mark these failing tests with "continue-on-error: true" flag.
Once we add the features and publish experimental languages, we can revert the "continue-on-error" flag and make these tests mandatory.

Usually, as the tests take so long, the maintainers prefer not to run them and save the time.
Today there is only one test of that sort: "SWIGLANG: mzscheme"

If you improve the action: use artefacts and copy the tools from one job to another, than we can save time and add more tests :-)

LecrisUT commented:
> You can put the 'lang' and 'tool' after as most developers know TCL is a language and 'GCC' a compiler (tool)
>
> As for emoji, I like the idea. You're choosing is better than my ideas.

Thanks for the feedback, I'll try out some simplifications along those lines.

> What we can do is mark these failing tests with "continue-on-error: true" flag.
> Once we add the features and publish experimental languages, we can revert the "continue-on-error" flag and make these tests mandatory.

That is already what is done there (`experimental == continue-on-error`)

> There is no difference.
> If the PR fails, it will **not** be merged into 'main'.
> Developers can not differ the 2 kinds.

Indeed, that is why I was trying to get the attention of the maintainers to discuss this change. Basically on my projects I set up a required workflow that points to `Pass` job so that the red jobs are ignored. [Here](https://github.com/re-actors/alls-green/pull/27) are my notes on how to mask the results, and you can find it in action [here](https://github.com/LecrisUT/CMake-Template) (also in the PRs there).

I can disable the `experimental` jobs, but do you have a suggestion on where to track that these need to be re-enabled?

erezgeva commented:
> > You can put the 'lang' and 'tool' after as most developers know TCL is a language and 'GCC' a compiler (tool)
> > As for emoji, I like the idea. You're choosing is better than my ideas.
> 
> Thanks for the feedback, I'll try out some simplifications along those lines.
> 
> > What we can do is mark these failing tests with "continue-on-error: true" flag.
> > Once we add the features and publish experimental languages, we can revert the "continue-on-error" flag and make these tests mandatory.
> 
> That is already what is done there (`experimental == continue-on-error`)
> 
> > There is no difference.
> > If the PR fails, it will **not** be merged into 'main'.
> > Developers can not differ the 2 kinds.
> 
> Indeed, that is why I was trying to get the attention of the maintainers to discuss this change. Basically on my projects I set up a required workflow that points to `Pass` job so that the red jobs are ignored. [Here](https://github.com/re-actors/alls-green/pull/27) are my notes on how to mask the results, and you can find it in action [here](https://github.com/LecrisUT/CMake-Template) (also in the PRs there).
> 
> I can disable the `experimental` jobs, but do you have a suggestion on where to track that these need to be re-enabled?

Seems you look for a project decision. 
I am merely a Developer happy to contribute and help.
You should talk to @ojwb about this issue.

erezgeva reviewed (approved):
I left my feedback.
The decision to have failed tests is up to the project Admins.
